### PR TITLE
scale down observer

### PR DESCRIPTION
**Pull request by catpineapple**

### CODE REVIEW

File: pkg/controller/sub_controller/fe/statefulset.go [hunk 1]
```diff
@@ -1,13 +1,73 @@
...
+	if err != nil {
+		klog.Errorf("ScaleOutObserver get service failed, service:%s, namespace:%s, err:%s", serviceName, targetStatefulSet.Namespace, err.Error())
+		return err
+	}
+
+	var queryPort string
+	for _, port := range service.Spec.Ports {
+		if port.Name == resource.GetPortKey(resource.QUERY_PORT) {
+			queryPort = strconv.FormatInt(int64(port.Port), 10)
+			continue
```
Comment (intelligentfu) [inline pkg/controller/sub_controller/fe/statefulset.go]: break

File: pkg/controller/sub_controller/fe/controller.go [hunk 2]
```diff
@@ -107,29 +114,86 @@ func (fc *Controller) Sync(ctx context.Context, cluster *v1.DorisCluster) error
...
+		}
+	}
+
 	return nil
 }
 
-func min(a, b int32) int32 {
-	if a < b {
-		return a
+func (fc *Controller) operator(ctx context.Context, st appv1.StatefulSet, cluster v1.DorisCluster) error {
```
Comment (intelligentfu) [inline pkg/controller/sub_controller/fe/controller.go]: statefulset should transfer pointer.

File: pkg/controller/doriscluster_controller.go [hunk 3]
```diff
@@ -125,12 +127,23 @@ func (r *DorisClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request
...
 		if err := rc.UpdateComponentStatus(dcr); err != nil {
 			klog.Errorf("DorisClusterReconciler reconcile update component %s status failed.err=%s\n", rc.GetControllerName(), err.Error())
 			return requeueIfError(err)
 		}
 	}
 
+	status := true
+	for _, rc := range r.Scs {
+		//check component status to cluster.
+		status = status && rc.GetComponentStatus(dcr) == dorisv1.Available
```
Comment (intelligentfu) [inline pkg/controller/doriscluster_controller.go]: status always is true
if rc.GetComponentStatus(dcr) != dorisv1.Available {
       status = false
}

File: pkg/common/utils/k8s/client.go [hunk 4]
```diff
@@ -223,3 +225,70 @@ func GetConfigMaps(ctx context.Context, k8scient client.Client, namespace string
...
+// get the Service by namespace and name.
+func GetService(ctx context.Context, k8sclient client.Client, namespace, name string) (*corev1.Service, error) {
+	var svc corev1.Service
+	if err := k8sclient.Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, &svc); err != nil {
+		return nil, err
+	}
+	return &svc, nil
+}
+
+func GetPods(ctx context.Context, k8sclient client.Client, namespace string) (corev1.PodList, error) {
```
Comment (intelligentfu) [inline pkg/common/utils/k8s/client.go]: can't only use namespace

File: pkg/common/utils/k8s/client.go [hunk 5]
```diff
@@ -223,3 +225,70 @@ func GetConfigMaps(ctx context.Context, k8scient client.Client, namespace string
...
+		return pods, err
+	}
+
+	for _, pod := range pods.Items {
+		fmt.Printf("pod --- Name: %s,  pod: %s \n", pod.GetName(), pod.Status.PodIP)
+	}
+	return pods, nil
+}
+
+func GetConfig(ctx context.Context, k8sclient client.Client, configMapInfo *dorisv1.ConfigMapInfo, namespace string, componentType dorisv1.ComponentType) (map[string]interface{}, error) {
```
Comment (intelligentfu) [inline pkg/common/utils/k8s/client.go:256]: getStartConfig.
add comment, add  comment, add comment.
this function has complicated logic, please add a comment.

File: pkg/common/utils/mysql/doris.go [hunk 6]
```diff
@@ -54,3 +60,23 @@ type Backend struct {
 	HeartbeatFailureCounter int    `json:"heartbeat_failure_counter" db:"HeartbeatFailureCounter"`
 	NodeRole                string `json:"node_role" db:"NodeRole"`
 }
+
+func SortAndSliceObserves(nodes []Frontend, scaleNumber int) []Frontend {
```
Comment (intelligentfu) [inline pkg/common/utils/mysql/doris.go]: add comment,

File: pkg/common/utils/mysql/mysql.go [hunk 7]
```diff
@@ -25,17 +24,21 @@ func NewDorisSqlDB(cfg DBConfig) (*DB, error) {
 	dsn := fmt.Sprintf("%s:%s@tcp(%s:%s)/%s", cfg.User, cfg.Password, cfg.Host, cfg.Port, cfg.Database)
 	db, err := sqlx.Open("mysql", dsn)
 	if err != nil {
-		klog.Errorf("failed open doris sql client connection, err: %s \n", err)
+		klog.Errorf("failed open doris sql client connection, err: %s \n", err.Error())
```
Comment (intelligentfu) [inline pkg/common/utils/mysql/mysql.go]: log format

File: pkg/common/utils/mysql/mysql.go [hunk 8]
```diff
@@ -86,21 +89,43 @@ func (db *DB) DropObserver(nodes []Frontend) error {
...
 		alter = alter + fmt.Sprintf(`ALTER SYSTEM DROP OBSERVER "%s:%d";`, node.Host, node.EditLogPort)
 	}
+	klog.Errorf("drop sql is :%s", alter)
+	//return errors.New("print err")
 	_, err := db.Exec(alter)
 	return err
 }
 
-func (db *DB) GetMaster() (*Frontend, error) {
+func (db *DB) GetObserves() (*[]Frontend, error) {
```
Comment (intelligentfu) [inline pkg/common/utils/mysql/mysql.go]: the name spelling error

File: pkg/common/utils/resource/configmap.go [hunk 9]
```diff
@@ -51,6 +53,21 @@ var defMap = map[string]int32{
 	BROKER_IPC_PORT:        8000,
 }
 
+func IsFQDN(config map[string]interface{}) bool {
+	// not use configmap
```
Comment (intelligentfu) [inline pkg/common/utils/resource/configmap.go:60]: get the start mode, not only fqdn. should classfy different start type.

**Review by intelligentfu — COMMENTED** [review]
check

File: pkg/common/utils/mysql/mysql.go [hunk 10]
```diff
@@ -86,21 +89,43 @@ func (db *DB) DropObserver(nodes []Frontend) error {
...
+	for _, fe := range frontends {
+		if fe.Role == FE_OBSERVE_ROLE {
+			res = append(res, fe)
+		}
+	}
+	return &res, nil
+}
+
+// GetFollowers return fe master,all followers(including master) and err
+func (db *DB) GetFollowers() (*Frontend, *[]Frontend, error) {
```
Comment (intelligentfu) [inline pkg/common/utils/mysql/mysql.go]: *Frontend, []*Frontend, *[]Frontends differenet.

File: pkg/controller/doriscluster_controller.go (hunk 3 above)
Comment (intelligentfu) [inline pkg/controller/doriscluster_controller.go]: status = (rc.GetComponentStatus(dcr) == dorisv1.Available) && status

File: api/doris/v1/doriscluster_webhook.go [hunk 11]
```diff
@@ -56,8 +58,35 @@ func (r *DorisCluster) ValidateCreate() error {
...
 func (r *DorisCluster) ValidateUpdate(old runtime.Object) error {
 	klog.Info("validate update", "name", r.Name)
+	var errors []error
+	oldCluster := old.(*DorisCluster)
+
+	//operableErr := checkClusterOperable(r)
+
+	// fe follower not allowed scale down
+	if *oldCluster.Spec.FeSpec.ElectionNumber > *r.Spec.FeSpec.ElectionNumber {
+		errors = append(errors, fmt.Errorf("scale down in the number of ElectionNumber are not allowed"))
```
Comment (intelligentfu) [inline api/doris/v1/doriscluster_webhook.go]: ReasonCode， Reason,  messsage

File: pkg/controller/sub_controller/fe/statefulset.go [hunk 12]
```diff
@@ -1,13 +1,152 @@
...
+
+}
+
+// buildFrontendMap
+// input ipMap key is podIP,value is fe.podName(from 'kubectl get pods -owide')
+// return frontendMap key is fe index ,value is frontend
+func buildFrontendMap(frontends []mysql.Frontend, ipMap map[string]string, podNamePre string) (map[int]mysql.Frontend, error) {
+	frontendMap := make(map[int]mysql.Frontend)
+	for _, fe := range frontends {
+		var podName string
```
Comment (intelligentfu) [inline pkg/controller/sub_controller/fe/statefulset.go]: array:=split(name, ".")
barray:=split(array[0],"-")
seqNumber := barray[len(barray)-1]

File: pkg/controller/sub_controller/fe/statefulset.go [hunk 13]
```diff
@@ -1,13 +1,152 @@
...
+		if strings.HasPrefix(fe.Host, podNamePre) {
+			// use fqdn, not need ipMap
```
Comment (intelligentfu) [inline pkg/controller/sub_controller/fe/statefulset.go]: 通过 clustername split， clustername,xxxx
split(xxxx, "-")

File: pkg/controller/sub_controller/fe/statefulset.go [hunk 14]
```diff
@@ -1,13 +1,152 @@
...
+	// drop node and return
+	err = db.DropObserver(observes)
+	return err
+
+}
+
+// buildFrontendMap
+// input ipMap key is podIP,value is fe.podName(from 'kubectl get pods -owide')
+// return frontendMap key is fe index ,value is frontend
+func buildFrontendMap(frontends []mysql.Frontend, ipMap map[string]string, podNamePre string) (map[int]mysql.Frontend, error) {
```
Comment (intelligentfu) [inline pkg/controller/sub_controller/fe/statefulset.go]: buildSeqNumberToFrontend

File: pkg/controller/sub_controller/fe/statefulset.go [hunk 15]
```diff
@@ -1,13 +1,152 @@
...
+	fmt.Printf("queryPort : %d \n", queryPort)
+	// connect to doris sql
+	dbConf := mysql.DBConfig{
+		User:     adminUserName,
+		Password: password,
+		Host:     serviceName,
+		Port:     strconv.FormatInt(int64(queryPort), 10),
+		Database: "mysql",
+	}
+	klog.Infof("dbConf %+v", dbConf)
```
Comment (intelligentfu) [inline pkg/controller/sub_controller/fe/statefulset.go]: debug + TODO:

File: pkg/controller/sub_controller/fe/statefulset.go [hunk 16]
```diff
@@ -1,13 +1,152 @@
...
 	st := resource.NewStatefulSet(dcr, v1.Component_FE)
 	st.Spec.Template = fc.buildFEPodTemplateSpec(dcr)
 	return st
 }
+
+// ScaleDownObserver handles doris'SQL(drop frontend) through the MySQL client when dealing with scale in observer
+// targetStatefulSet is new StatefulSet
+// targetDCR is new dcr
+// scaleNumber is the number of Observer needing scale in
+func (fc *Controller) ScaleDownObserver(
```
Comment (intelligentfu) [inline pkg/controller/sub_controller/fe/statefulset.go]: DropObserverFromMaster
DropObserver

File: api/doris/v1/types.go [hunk 17]
```diff
@@ -312,6 +312,36 @@ type DorisServicePort struct {
...
+	RETRY_OPERATOR_FE     RetrySign = "fe"
+	RETRY_OPERATOR_BE     RetrySign = "be"
+	RETRY_OPERATOR_CN     RetrySign = "cn"
+	RETRY_OPERATOR_BROKER RetrySign = "broker"
+	RETRY_OPERATOR_NO     RetrySign = ""
+)
+
+type ClusterSituation struct {
+	//describe the current status of the doris cluster, record operable, scaling, upgrading and restarting.
+	Situation Situation `json:"situation,omitempty"`
```
Comment (intelligentfu) [inline api/doris/v1/types.go]: reference status in pod , statefulset, etc

File: pkg/common/utils/k8s/client.go [hunk 18]
```diff
@@ -223,3 +225,70 @@ func GetConfigMaps(ctx context.Context, k8scient client.Client, namespace string
...
+	if err := k8sclient.Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, &svc); err != nil {
+		return nil, err
+	}
+	return &svc, nil
+}
+
+func GetPods(ctx context.Context, k8sclient client.Client, namespace string) (corev1.PodList, error) {
+	pods := corev1.PodList{}
+
+	options := client.ListOptions{
```
Comment (intelligentfu) [inline pkg/common/utils/k8s/client.go]: add labels

File: pkg/common/utils/k8s/client.go [hunk 19]
```diff
@@ -223,3 +225,70 @@ func GetConfigMaps(ctx context.Context, k8scient client.Client, namespace string
...
+		Namespace: namespace,
+	}
+
+	err := k8sclient.List(ctx, &pods, &options)
+	if err != nil {
+		return pods, err
+	}
+
+	for _, pod := range pods.Items {
```
Comment (intelligentfu) [inline pkg/common/utils/k8s/client.go]: this code should not be committed.

**Review by intelligentfu — COMMENTED** [review]
check

**Comment by catpineapple** [comment]
not  close pvc resource

File: api/doris/v1/doriscluster_util.go [hunk 20]
```diff
@@ -331,3 +332,28 @@ func getFeAddrForBroker(dcr *DorisCluster) (string, int) {
...
+}
+
+// SetPhaseOperable set cluster status is operable and retry sign is empty
+func SetPhaseOperable(dcr *DorisCluster) {
+	dcr.Status.ClusterPhase.Phase = PHASE_OPERABLE
+	dcr.Status.ClusterPhase.Retry = RETRY_OPERATOR_NO
+}
+
+// SetPhaseInit if Phase is ""
+func SetPhaseInit(dcr *DorisCluster) {
```
Comment (intelligentfu) [inline api/doris/v1/doriscluster_util.go]: not need a function.

File: api/doris/v1/doriscluster_util.go [hunk 21]
```diff
@@ -331,3 +332,28 @@ func getFeAddrForBroker(dcr *DorisCluster) (string, int) {
...
+	if dcr.Spec.AdminUser != nil {
+		return dcr.Spec.AdminUser.Name, dcr.Spec.AdminUser.Password
+	}
+	return "root", ""
+}
+
+// SetPhaseOperable set cluster status is operable and retry sign is empty
+func SetPhaseOperable(dcr *DorisCluster) {
+	dcr.Status.ClusterPhase.Phase = PHASE_OPERABLE
+	dcr.Status.ClusterPhase.Retry = RETRY_OPERATOR_NO
```
Comment (intelligentfu) [inline api/doris/v1/doriscluster_util.go]: retry is operator concern should not appear in status

File: api/doris/v1/doriscluster_webhook.go [hunk 22]
```diff
@@ -56,8 +58,35 @@ func (r *DorisCluster) ValidateCreate() error {
 // ValidateUpdate implements webhook.Validator so a unnamedwatches will be registered for the type
 func (r *DorisCluster) ValidateUpdate(old runtime.Object) error {
 	klog.Info("validate update", "name", r.Name)
+	var errors []error
+	oldCluster := old.(*DorisCluster)
+
+	//operableErr := checkClusterOperable(r)
+
+	// fe follower not allowed scale down
+	if *oldCluster.Spec.FeSpec.ElectionNumber > *r.Spec.FeSpec.ElectionNumber {
```
Comment (intelligentfu) [inline api/doris/v1/doriscluster_webhook.go]: this should allowed, if not, follower will not scale down in any way.

File: api/doris/v1/doriscluster_webhook.go [hunk 23]
```diff
@@ -56,8 +58,35 @@ func (r *DorisCluster) ValidateCreate() error {
...
+		errors = append(errors, fmt.Errorf("'FeSpec.ElectionNumber' error: scale down in the number of ElectionNumber are not allowed"))
+	}
+
+	// fe FeSpec.Replicas must greater than or equal to FeSpec.ElectionNumber
+	if *r.Spec.FeSpec.Replicas < *r.Spec.FeSpec.ElectionNumber {
+		errors = append(errors, fmt.Errorf("'FeSpec.Replicas' error: changes in the number of FeSpec.Replicas must greater than or equal to FeSpec.ElectionNumber"))
```
Comment (intelligentfu) [inline api/doris/v1/doriscluster_webhook.go]: fe replicas should greater than or equal electionNumber.

File: api/doris/v1/doriscluster_webhook.go [hunk 24]
```diff
@@ -56,8 +58,35 @@ func (r *DorisCluster) ValidateCreate() error {
...
+	}
+
+	if oldCluster.Status.ClusterPhase.Phase != PHASE_OPERABLE && oldCluster.Status.ClusterPhase.Retry != RETRY_OPERATOR_FE && *r.Spec.FeSpec.Replicas != *oldCluster.Spec.FeSpec.Replicas {
+		errors = append(errors, fmt.Errorf("ClusterOperationalConflicts error: there is a conflict in CRD modify. currently, cluster Phase is %+v ", oldCluster.Status.ClusterPhase))
+	}
+
+	if oldCluster.Status.ClusterPhase.Phase != PHASE_OPERABLE &&
+		oldCluster.Status.ClusterPhase.Retry != RETRY_OPERATOR_BE &&
+		*r.Spec.BeSpec.Replicas != *oldCluster.Spec.BeSpec.Replicas {
+		errors = append(errors, fmt.Errorf("there is a conflict in CRD modify. currently, cluster Phase is %+v ", oldCluster.Status.ClusterPhase))
```
Comment (intelligentfu) [inline api/doris/v1/doriscluster_webhook.go]: don't use %v use %s

File: api/doris/v1/types.go [hunk 25]
```diff
@@ -312,6 +312,36 @@ type DorisServicePort struct {
 	TargetPort int32 `json:"targetPort,omitempty"`
 }
 
+// Cluster operation and maintenance status
+// The cluster can only be adjusted in the operable state
+type Phase string
+
+const (
+	PHASE_INITIALIZING Phase = "initializing"
+	PHASE_OPERABLE     Phase = "operable"
```
Comment (intelligentfu) [inline api/doris/v1/types.go]: not phase name as `operable`

File: api/doris/v1/types.go [hunk 26]
```diff
@@ -386,6 +419,7 @@ const (
 // +kubebuilder:subresource:status
 // +kubebuilder:resource:shortName=dcr
 // +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="ClusterPhase",type=string,JSONPath=`.status.clusterPhase.phase`
 // +kubebuilder:printcolumn:name="FeStatus",type=string,JSONPath=`.status.feStatus.componentCondition.phase`
```
Comment (intelligentfu) [inline api/doris/v1/types.go]: not needed.

File: pkg/common/utils/k8s/client.go [hunk 27]
```diff
@@ -223,3 +225,82 @@ func GetConfigMaps(ctx context.Context, k8scient client.Client, namespace string
...
+// get the Service by namespace and name.
+func GetService(ctx context.Context, k8sclient client.Client, namespace, name string) (*corev1.Service, error) {
+	var svc corev1.Service
+	if err := k8sclient.Get(ctx, types.NamespacedName{Name: name, Namespace: namespace}, &svc); err != nil {
+		return nil, err
+	}
+	return &svc, nil
+}
+
+func GetPods(ctx context.Context, k8sclient client.Client, targetDCR dorisv1.DorisCluster) (corev1.PodList, error) {
```
Comment (intelligentfu) [inline pkg/common/utils/k8s/client.go]: GetPods is a common function, so why use fe filter in the function?

File: pkg/common/utils/k8s/client.go [hunk 28]
```diff
@@ -223,3 +225,82 @@ func GetConfigMaps(ctx context.Context, k8scient client.Client, namespace string
...
+	pods := corev1.PodList{}
+
+	err := k8sclient.List(ctx, &pods, client.InNamespace(targetDCR.Namespace), client.MatchingLabels(dorisv1.GetPodLabels(&targetDCR, dorisv1.Component_FE)))
+	if err != nil {
+		return pods, err
+	}
+
+	for _, pod := range pods.Items {
+		fmt.Printf("pod --- Name: %s,  pod: %s \n", pod.GetName(), pod.Status.PodIP)
```
Comment (intelligentfu) [inline pkg/common/utils/k8s/client.go]: this log format should not appear.

File: pkg/controller/sub_controller/fe/sql_client.go [hunk 29]
```diff
@@ -0,0 +1,198 @@
...
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sort"
+	"strconv"
+	"strings"
+)
+
+// ControlClusterPhaseAndPreOperation means Pre-operation and status control on the client side
+func (fc *Controller) controlClusterPhaseAndPreOperation(ctx context.Context, cluster *v1.DorisCluster) error {
```
Comment (intelligentfu) [inline pkg/controller/sub_controller/fe/sql_client.go:22]: this function is useless.

File: pkg/controller/sub_controller/sub_controller.go [hunk 30]
```diff
@@ -327,3 +331,120 @@ func (d *SubDefaultController) patchPVCs(ctx context.Context, dcr *dorisv1.Doris
...
+// recycleFEResources pvc resource for fe recycle
+func (d *SubDefaultController) recycleFEResources(ctx context.Context, dcr *dorisv1.DorisCluster) error {
+	if len(dcr.Spec.FeSpec.PersistentVolumes) != 0 {
+		return d.listPersistentVolumeClaimForDelete(ctx, dcr, dorisv1.Component_FE)
+	}
+	return nil
+}
+
+// recycleBEResources pvc resource for be recycle
+func (d *SubDefaultController) recycleBEResources(ctx context.Context, dcr *dorisv1.DorisCluster) error {
```
Comment (intelligentfu) [inline pkg/controller/sub_controller/sub_controller.go]: this is very very dangerous!!!!!! I have told you don't clear be resources.

**Review by intelligentfu — COMMENTED** [review]
check

File: pkg/controller/sub_controller/fe/sql_client.go [hunk 31]
```diff
@@ -0,0 +1,198 @@
...
+				return err
+			}
+		}
+		if scaleNumber < 0 {
+			if err := fc.dropObserverFromSqlClient(ctx, fc.K8sclient, cluster); err != nil {
+				klog.Errorf("ScaleDownObserver failed, err:%s ", err.Error())
+				return err
+			}
+			if scaleNumber != -1 {
+				klog.Info("controlClusterPhaseAndPreOperation scale down observer task is not completed, %d tasks are left. ", -scaleNumber-1)
```
Comment (intelligentfu) [inline pkg/controller/sub_controller/fe/sql_client.go]: 出现了负号？？？？？

File: pkg/controller/sub_controller/sub_controller.go [hunk 32]
```diff
@@ -327,3 +331,120 @@ func (d *SubDefaultController) patchPVCs(ctx context.Context, dcr *dorisv1.Doris
...
+
+// deletePVCs will Loop to remove excess pvc
+func (d *SubDefaultController) deletePVCs(ctx context.Context, dcr *dorisv1.DorisCluster, selector map[string]string,
+	pvcs []corev1.PersistentVolumeClaim, stsName string, volume dorisv1.PersistentVolume, replicas int32) error {
+	baseOrdinal := len(pvcs)
+
+	var mergeError error
+	for ; baseOrdinal > int(replicas); baseOrdinal-- {
+		pvc := resource.BuildPVC(volume, selector, dcr.Namespace, stsName, strconv.Itoa(baseOrdinal-1))
+		if err := k8s.DeletePVC(ctx, d.K8sclient, dcr.Namespace, pvc.Name); err != nil {
```
Comment (intelligentfu) [inline pkg/controller/sub_controller/sub_controller.go]: not right, first get next delete.

**Review by intelligentfu — COMMENTED** [review]
check

**Review by intelligentfu — COMMENTED** [review]
check

File: pkg/controller/sub_controller/fe/sql_client.go [hunk 33]
```diff
@@ -0,0 +1,198 @@
...
+			klog.Errorf("buildSeqNumberToFrontend can not split pod name,pod name: %s,err:%s", podSignName, err.Error())
+			return nil, err
+		}
+		frontendMap[num] = fe
+	}
+	return frontendMap, nil
+}
+
+// sort fe by index and return top scaleNumber
+func getTopFrontends(frontendMap map[int]*mysql.Frontend, scaleNumber int32) []*mysql.Frontend {
```
Comment (intelligentfu) [inline pkg/controller/sub_controller/fe/sql_client.go]: findNeedDeleteFrontends

File: pkg/controller/sub_controller/sub_controller.go [hunk 34]
```diff
@@ -327,3 +331,120 @@ func (d *SubDefaultController) patchPVCs(ctx context.Context, dcr *dorisv1.Doris
...
+		}
+		if err := d.deletePVCs(ctx, dcr, selector, claims, stsName, volume, replicas); err != nil {
+			mergeError = utils.MergeError(mergeError, err)
+		}
+	}
+	return mergeError
+}
+
+// deletePVCs will Loop to remove excess pvc
+func (d *SubDefaultController) deletePVCs(ctx context.Context, dcr *dorisv1.DorisCluster, selector map[string]string,
```
Comment (intelligentfu) [inline pkg/controller/sub_controller/sub_controller.go:408]: where used pvcs?

File: api/doris/v1/types.go [hunk 35]
```diff
@@ -312,6 +312,36 @@ type DorisServicePort struct {
...
+	RETRY_OPERATOR_BROKER RetrySign = "broker"
+	RETRY_OPERATOR_NO     RetrySign = ""
+)
+
+type ClusterPhase struct {
+	//describe the current status of the doris cluster, record operable, scaling, upgrading and restarting.
+	Phase Phase `json:"phase,omitempty"`
+
+	// describe change status task of the doris cluster need retry
+	Retry RetrySign `json:"retry,omitempty"`
```
Comment (intelligentfu) [inline api/doris/v1/types.go]: cluster spec =每一个服务的spec，每一个pod的状态满足预期。

File: pkg/controller/sub_controller/fe/sql_client.go [hunk 36]
```diff
@@ -0,0 +1,198 @@
...
+		// means other task running, send Event warning
+		fc.K8srecorder.Eventf(
+			cluster, sub_controller.EventWarning,
+			sub_controller.ClusterOperationalConflicts,
+			"There is a conflict in crd operation. currently, cluster Phase is %+v ", phase.Phase,
+		)
+		return errors.New(fmt.Sprintf("There is a conflict in crd operation. currently, cluster Phase is %+v ", phase.Phase))
+	}
+
+	if phase.Phase == v1.PHASE_OPERABLE || phase.Retry == v1.RETRY_OPERATOR_FE {
```
Comment (intelligentfu) [inline pkg/controller/sub_controller/fe/sql_client.go]: statefulset的pod的数量和所有pod的状态。

File: pkg/controller/sub_controller/fe/sql_client.go (hunk 29 above)
Comment (intelligentfu) [inline pkg/controller/sub_controller/fe/sql_client.go]: 判断是否可以执行 缩容
1. statefulset的replicas 和 pod 数量。
2. 所有pod的状态。

File: pkg/controller/sub_controller/sub_controller.go [hunk 37]
```diff
@@ -327,3 +331,120 @@ func (d *SubDefaultController) patchPVCs(ctx context.Context, dcr *dorisv1.Doris
 
 	return prepared
 }
+
+// RecycleResources pvc resource for recycle
+func (d *SubDefaultController) RecycleResources(ctx context.Context, dcr *dorisv1.DorisCluster, componentType dorisv1.ComponentType) error {
+	switch componentType {
+	case dorisv1.Component_FE:
+		return d.recycleFEResources(ctx, dcr)
+	case dorisv1.Component_BE:
```
Comment (intelligentfu) [inline pkg/controller/sub_controller/sub_controller.go]: case fe:
xxxx
default:
return

**Review by intelligentfu — APPROVED** [review]

**Pull request merged**